pydantic/_internal/_typing_extra.py

```diff
@@ -225,17 +225,54 @@ def parent_frame_namespace(*, parent_depth: int = 2, force: bool = False) -> dic
     return frame.f_locals
 
 
+def _type_convert(arg: Any) -> Any:
+    """Convert `None` to `NoneType` and strings to `ForwardRef` instances.
+
+    This is a backport of the private `typing._type_convert` function. When
+    evaluating a type, `ForwardRef._evaluate` ends up being called, and is
+    responsible for making this conversion. However, we still have to apply
+    it for the first argument passed to our type evaluation functions, similarly
+    to the `typing.get_type_hints` function.
+    """
+    if arg is None:
+        return NoneType
+    if isinstance(arg, str):
+        # Like `typing.get_type_hints`, assume the arg can be in any context,
+        # hence the proper `is_argument` and `is_class` args:
+        return _make_forward_ref(arg, is_argument=False, is_class=True)
+    return arg
+
+
+@overload
```

Comment:
I'm usually not a fan of having overloads with a different return value. It is a sign that this function should actually be split in two. I spent some (too much in fact) time figuring out a good API, but couldn't find a clean way to do so.

Reply:
Would this be easier if we just returned `dict[str, tuple[Any, bool]]` always, so when `lenient` is `False`, we just return `True` as the second part of the tuple?

Reply:
I wanted to do this as well, but then we loose the `NameError` exceptions, which are converted to `PydanticUndefinedAnnotation` exceptions during schema building :/